An alternative way to load peripherals information (#8)
eclipse-cdt-cloud/vscode-peripheral-inspector: docs/extending-svd-viewer.md

> # Extending SVD Viewer

Reviewer: ```suggestion
# Extending Peripheral Inspector
```

Also, should we rename this file?

> It is possible to extend the SVD Viewer and register new file extension providers with your VSCode extension. This method will provide reading non-svd formats and load the peripherals information in SVD Viewer. 

Reviewer: ```suggestion
It is possible to extend the Peripheral Inspector and register new file extension providers with your VSCode extension. This method will provide reading non-svd formats and load the peripherals information in Peripheral Inspector. 
```

> ## Building your VSCode Extension to extend SVD Viewer

Reviewer: ```suggestion
## Building your VSCode Extension to extend Peripheral Inspector
```

> This is a guide about how you can register new peripheral providers to SVD Viewer in your VSCode extension. Please refer to [VSCode Extension API](https://code.visualstudio.com/api) for more information about developing VSCode extensions.

Reviewer: ```suggestion
This is a guide about how you can register new peripheral providers to Peripheral Inspector in your VSCode extension. Please refer to [VSCode Extension API](https://code.visualstudio.com/api) for more information about developing VSCode extensions.
```

### Adding peripheral-viewer to your VSCode extension

> You need to install eclipse-cdt-cloud/vscode-svd-viewer to access the types information. You can use `npm` or `yarn` with the following arguments described below:

Reviewer: ```suggestion
You need to install eclipse-cdt-cloud/vscode-peripheral-inspector to access the types information. You can use `npm` or `yarn` with the following arguments described below:
```

Using with npm:
```bash

> npm install github:eclipse-cdt-cloud/vscode-svd-viewer

Reviewer: ```suggestion
npm install github:eclipse-cdt-cloud/vscode-peripheral-inspector
```

```
Using with yarn:
```bash

> yarn add github:eclipse-cdt-cloud/vscode-svd-viewer

Reviewer: ```suggestion
yarn add github:eclipse-cdt-cloud/vscode-peripheral-inspector
```

```

### Developing your extension

> To provide the peripherals information to SVD Viewer on debug session time, you need register your command which is going to construct the peripherals information. The command will receive `DebugSession` object as an input parameter and expects to return array of type `PeripheralOptions[]`.

Reviewer: ```suggestion
To provide the peripherals information to Peripheral Inspector on debug session time, you need register your command which is going to construct the peripherals information. The command will receive `DebugSession` object as an input parameter and expects to return array of type `PeripheralOptions[]`.
```

You can find the example command implementation below:

```js
import { ExtensionContext } from 'vscode';

> import { api } from "svd-viewer";

Reviewer: ```suggestion
import { api } from "peripheral-inspector";
```

# Extending SVD Viewer

It is possible to extend the SVD Viewer and register new file extension providers with your VSCode extension. This method will provide reading non-svd formats and load the peripherals information in SVD Viewer. 

## Building your VSCode Extension to extend SVD Viewer

This is a guide about how you can register new peripheral providers to SVD Viewer in your VSCode extension. Please refer to [VSCode Extension API](https://code.visualstudio.com/api) for more information about developing VSCode extensions.

### Adding peripheral-viewer to your VSCode extension

> You need to install eclipse-cdt-cloud/vscode-svd-viewer to access the types information. You can use `npm` or `yarn` with the following arguments described below:

Reviewer: A better approach would be to publish a small package containing the extension API export which can be consumed, would be a bit more work though.

This is what Arm do for the cmsis extension:

https://www.npmjs.com/package/@arm-software/vscode-cmsis-csolution

Using with npm:
```bash
npm install github:eclipse-cdt-cloud/vscode-svd-viewer
```
Using with yarn:
```bash
yarn add github:eclipse-cdt-cloud/vscode-svd-viewer
```

### Developing your extension

> To provide the peripherals information to SVD Viewer on debug session time, you need register your command which is going construct the peripherals information. The command will receive `DebugSession` object as an input parameter and expects to return array of type `PeripheralOptions[]`.

Reviewer: ```suggestion
To provide the peripherals information to SVD Viewer on debug session time, you need register your command which is going to construct the peripherals information. The command will receive `DebugSession` object as an input parameter and expects to return array of type `PeripheralOptions[]`.
```

Author: It is changed.

You can find the example command implementation below:

```js
import { ExtensionContext } from 'vscode';

> import { 

Reviewer: Concrete implementations aren't needed, use `type` to ensure none of the extension is included at compile time.

```suggestion
import type { 
```

Author: AccessType is an enum so we are unable to define it as `import type` :(

	PeripheralOptions, 
	PeripheralRegisterOptions, 
	ClusterOptions, 
	FieldOptions, 
	AccessType,
	IGetPeripheralsArguments,

>     IPeripheralsProvider

Reviewer: ```suggestion
        IPeripheralsProvider
```

Author: Changed.

} from "svd-viewer";

class MyExtensionProvider implements IPeripheralsProvider {
    public async getPeripherals (data: string, options: IGetPeripheralsArguments): Promise<PeripheralOptions[]> {
        // Load your peripherals data
		const peripherals: PeripheralOptions[] = ...
		return peripherals;
    }
}

export async function activate(context: ExtensionContext) {
	...
	// Get the eclipse-cdt.svd-viewer extension
    const svdViewerExtension = extensions.getExtension('eclipse-cdt.svd-viewer');

	// Invoke registerPeripheralsProvider method in eclipse-cdt.svd-viewer extension api
	// Register 'MyExtensionProvider' for files *.myext

>     svdViewerExtension?.exports?.registerPeripheralsProvider('myext', new MyExtensionProvider());

Reviewer: This could be more resilient (e.g. attach to an onActivation event)

	...
}
```

> You can check the type definitions (`PeripheralOptions`, `PeripheralRegisterOptions`, `ClusterOptions`, `FieldOptions`, `AccessType`) from [this document](./appendix-types.md).

Reviewer: Reference source file with extracted interfaces instead?
